### ISSUE TITLE
Update task version from the web interface

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-web#285]
EDIT: the most up-to-date description is in a comment below: https://github.com/fractal-analytics-platform/fractal-web/issues/285#issuecomment-1748430756

----

Here's a future feature wish:
Updating to a newer version of a task directly from the web interface for an existing workflow.

Currently, this can be done manually by 
a) Recreating the workflow and setting all parameters again
b) Download a workflow, changing the version of the task in the json, optionally changing parameters that are new, uploading it again.

The vision for this in the web interface would be that it replaces the task with a new version of the task and all the parameters pre-filled. If some parameters of the task changed, give some warning:
- If parameters were added, just list the new parameters
- If some parameters were removed, list the removed parameters and the values that the task had before updating them

This is not urgent, but will be useful when we update tasks versions with patch releases more :)

**Comment by tcompa** [comment]
Here's a first proposal, in view of ongoing discussions/meetings.

Task updating in web interface to new version.

- Goal: there is a workflow made of v0.11 tasks, and we want to fully migrate to v0.12. (question: should we also support migrating a mixed-version workflow? in principle also doable)
- Preliminary check 1: all workflow.task_list have a corresponding v0.12 (i.e. the tasks with corresponding source do exist).
_Non-goal_ (for the moment): user can specify a custom origin/target map (e.g. because source changed due to https://github.com/fractal-analytics-platform/fractal-server/issues/836)
- If checks 1 fails -> do not even start the procedure
- Preliminary check 2: check whether migration of old arguments is trivial for all tasks (that is, the old arguments also complywith the new task.args_schema JSON Schema).
- If check 2 succeeds, then start a loop over all WorkflowTask. For each task: remove it, and insert the new one (copying the old args and meta).
- If check 2 fails: to be defined. First proposal:
    - For each task that needs some action:
        - We show the error message and a text form to modify the JSON by hand
        - After user updates text and clicks "save", we parse it to JSON and validate with ajv again
        - We repeat this operation until arguments comply with the new JSON Schema.
    - After **all** task arguments have been fixed, we start the actual procedure of removing and re-inserting tasks.
- Safety measure: we should be able to go back to the original workflow, if needed. First, before starting any "remove and re-insert WorkflowTask" procedure, we export old workflow. Or maybe we copy it to a "MyWorkflow_backup_123132131" workflow. Then, when the procedure is successful, we can tell the user to go and remove the backup workflow (or we can do it by hand). If anything goes wrong, though, there is a simple way (automated or by hand) to go back to a clean state.

**Comment by tcompa** [comment]
OK, if the version update is a per-task operation (rather than per-workflow) then the procedure is even simpler, because checks 1 and 2 above should be performed a single time (vs "check N tasks and then decide how to proceed").

**Comment by tcompa** [comment]
Here is a recap, based on the discussion we had yesterday and on the previous comments in the issue.

## How to identify a task-update candidate

Given a task object (say it's `MyTask`), how can we identify whether a new version exists?

The naive way (which doesn't involve constructing and updating a global cache of tasks versions) is the following:
* Make sure that `MyTask.version` is defined. If not, stop this procedure.
* Fetch the full task list (`GET /api/v1/task/`)
* Discard all tasks that have name different from `MyTask.name` or owner different from `MyTask.owner`.
* Discard all tasks that do not have `MyTask.version` set.
* Discard all tasks with version less or equal to `MyTask.version` (see below for definition of "less or equal").

The tasks that are left in the list are all potential update candidates.

## How to compare tasks versions

We use the [`semver` module](https://www.npmjs.com/package/semver), e.g. as in https://github.com/fractal-analytics-platform/fractal-web/blob/main/src/lib/common/component_utilities.js#L11-L13:
```js
export function greatestVersionAsc(t1, t2) {
	const semverValidationOptions = {
		loose: true,
		includePrerelease: true
	};
	try {
		const t1Version = semver.valid(t1.version, semverValidationOptions);
		const t2Version = semver.valid(t2.version, semverValidationOptions);
		const t1VersionLt = semver.lte(t1Version, t2Version);
		if (t1VersionLt) return -1;
		if (!t1VersionLt) return 1;
	} catch {
		return 0;
	}
}
```

### A clarification (*not part of this issue*)

This approach works smoothly for standard release strings (`1`, `1.2`, `1.2.3`), while its behavior sometimes differs from the expected one for dev releases (e.g. `1.2.3a0` or `v0.1.dev27+g1458b59`). The main underlying reason is that SemVer is not fully compatible with PEP 440 (see e.g. https://github.com/semver/semver/issues/483). We are still able to parse PEP440 versions (see `loose: true` in the code snippet above), but the result may be unexpected.
Also: it's not fully clear where version strings like `v0.1.dev27+g1458b59` come from, but if they are correct then we should add them to https://github.com/fractal-analytics-platform/fractal-web/blob/main/__tests__/semver.test.js.

Let's review this as part of #286.


## Planned changes

1. In a workflow page, when we select one item from the task list we see the three tabs Arguments/Meta/Info. Let's add a new tab "Version" (or a similar name), which concerns the feature discussed in this issue.
2. Upon entering the Version tab, we identify the current task (which is the `task` property of the selected `WorkflowTask` object) and we look for task-update candidates. Note that this process should be stopped if the task attribute `args_schema` is `null` or `undefined` (at least in the first implementation), since in that case we would not be able to validate the new task arguments.
3. If there exist new version of the task, we list them in a dropdown menu and let the user select one of them (on the line of "New versions of this task exist: 1.2.3, 1.2.4, 1.2.5. Select one and proceed with the update").
4. We now have all the information we will need: two JSON Schemas (`oldTask.args_schema` and `newTask.args_schema`), and a data object `oldArgs` (which is the `args` property of the selected WorkflowTask object).
5. The goal is to transform `oldArgs` into some `newArgs` which is valid under `newTask.args_schema` (validation should always be performed with `ajv`, which we already use elsewhere). In some (most?) cases, just setting `newArgs` equal to `oldArgs` is enough, because the new schema is very similar to the previous one (e.g. we only added a non-required argument). In this case, we should tell the user that an automatic update can be performed, ask for confirmation, and proceed. "Proceeding" consists:
    * Make a copy of `oldArgs`, and of the `order` property of the current WorkflowTask (i.e. its positional index in the task list)
    * Remove the current WorkflowTask from the workflow, with something like `DELETE /api/v1/project/{project_id}/workflow/{workflow_id}/wftask/{old_workflow_task_id}`
    * Insert the new task into the workflow, with something like `POST /api/v1/project/{project_id}/workflow/{workflow_id}/wftask/?task_id={new_task_id}`. This also requires passing a payload like `{"args": newArgs, "meta": oldMeta, "order": order}` (where `order` and `meta` should be the same as in the old case).
6. In other cases, the migration may not be trivial (e.g. because we renamed a property of the schema, or we added a new required argument), and this is the trickiest case to handle. Here is my first naive proposal, but other suggestions are welcome:
    * We show a text prompt field, pre-filled with something like `JSON.stringify(oldArgs, null, 2)`.
    * The user can modify the text and click a "check" button, which parses the text to an object and then checks (via `ajv`) whether it complies with `newTask.args_schema`.
    * After iterating this procedure as many times as needed, if the user is able to produce a valid object then they can click "save" (this can also be merged with the "check" button..), and this assigns the current value of the text prompt to `newArgs`. The rest of the procedure is as in point 5 (removing a WorkflowTask and inserting a new Task)

### Unit tests

Some parts of this feature (e.g. the identification of new versions) are reasonably well defined and unit-testable. I think it would be nice to have a separate component, and a `vitest` test like:
```js
const task_list = [
    {name="My Task", owner=null, version="1.2.3", ...},
    {name="My Task", owner=null, version="1.2.4", ...},
    {name="My Task", owner=null, version="2", ...},
    {name="My Other Task", owner=null, version="1.2.3", ...},
    {name="My Other Task", owner="admin", version="1.3", ...}
    ]

let SomeTask = task_list[1];
// produce list of update candidates for SomeTask, and check that it has two items
let SomeTask = task_list[3];
// produce list of update candidates for SomeTask, and check that it is empty
```

And then I don't know whether tests are relevant also for the feature in point 6, depending on how it is implemented.

### Minor additional feature

If we anticipate the check of available updated versions, then we can also modify the the "Version"-tab button based on this information. For instance we could disable it if there is no new version. Or we could make it more prominent when a new version exists. This would be nice to have, but it's **not** necessary for a first version of this feature.

**Comment by jluethi** [comment]
Great summary @tcompa , I agree with the suggested implementation!

> If we anticipate the check of available updated versions, then we can also modify the the "Version"-tab button based on this information. For instance we could disable it if there is no new version. Or we could make it more prominent when a new version exists. This would be nice to have, but it's not necessary for a first version of this feature.

Agreed, this will be nice to have! We can start without this, test how well the version updating works. Once we're confident in its stability, we start pushing it more by highlighting it stronger when available :)
